Implement Authentication System with Custom User Model

najamulhassanarbi opened the pull request:
**Description**
This PR implements a comprehensive authentication system for the project. It includes user registration, login, logout, password reset functionality, and integrates a custom user model. The implementation also includes middleware for JWT-based authentication, a custom backend, and setup for PostgreSQL.

### Changes Made:

### > **Custom User Model:**

Implemented a custom user model (CustomUser) to extend the default Django user model with additional fields like role, avatar, bio, and timestamps.

### > **Forms:**

1. Created CustomUserCreationForm for user registration.
2. Created LoginForm for user login.
3. Created UserUpdateForm for updating user profile information.

### **> Views:**

1. SignUpView: Handles user registration.
2. CustomLoginView: Handles user login.
3. CustomLogoutView: Handles user logout.
4. ProfileView: Displays and updates the user’s profile.
5. ResetPasswordView: Handles password reset functionality.
6. HomeView: Displays the home page for logged-in users.

### > **Middleware and Custom Backend:**

1. Implemented JWT-based authentication middleware.
2. Created a custom authentication backend to support email-based login.

### > **Templates:**

1. Created templates for user registration, login, logout, profile update, password reset, and home page.
2. Included Bootstrap for styling the forms and pages.

### > **URL Configuration:**

1. Defined URL patterns for the authentication views, including signup, login, logout, profile update, and password reset.

### > **Database:**

1. Configured the project to use PostgreSQL.
2. Ensured the custom user model is properly integrated with PostgreSQL.

### > **Email Configuration:**

1. Set up email backend for password reset functionality.

najamulhassanarbi committed:
starter Project

najamulhassanarbi committed:
removed migrations files

najamulhassanarbi committed:
added custom authentication using jwt

najamulhassanarbi committed:
added authentication, reset password

Khuzaima-bashir commented on ArbiShop/users/forms.py:
```diff
@@ -0,0 +1,101 @@
+    class Meta:
+        model = User
+        fields = ['first_name', 'last_name', 'email', 'password1', 'password2']
+
+    # class Meta:
+    #     model = User
+    #     fields = ("email",)
+    #     widgets = {
+    #         'email': forms.EmailInput(attrs={'class': 'form-control', 'placeholder': 'Enter email'}),
+    #         'password1': forms.PasswordInput(attrs={'class': 'form-control', 'placeholder': 'Enter password'}),
+    #         'password2': forms.PasswordInput(attrs={'class': 'form-control', 'placeholder': 'Confirm password'}),
+    #     }
```
Remove it pep-8

Khuzaima-bashir commented on ArbiShop/users/managers.py:
```diff
@@ -0,0 +1,34 @@
+            raise ValueError(_("Superuser must have is_staff=True."))
+        if extra_fields.get("is_superuser") is not True:
+            raise ValueError(_("Superuser must have is_superuser=True."))
+        return self.create_user(email, password, **extra_fields)
```
line at end of file?

najamulhassanarbi committed:
it has to be merged. Mentor said it.
Merge remote-tracking branch 'origin/main' into ArbiShop

najamulhassanarbi committed:
product listing, filters for search, cart functionality (#5)

* product listing, filters for search, cart functionality

* refactored code

* added migrations file, infline forms for admins

najamulhassanarbi committed:
Integrate Stripe Payment Gateway and Jazzmin for Enhanced Admin Dashboard (#6)

* initial commit for order

* updated models

* added the stripe for payment, added jazzmin for admin dashboard

* added user profile, filters, slider, category bar, refactored

* added contstants.py file, removed prints, comments, added doc string

* added the config model, pagination is being fetched from database

najamulhassanarbi committed:
refactor price to pkr

najamulhassanarbi committed:
Implement Seller Dashboard Feature (#7)

* added the dashboard, for seller,included products CRUD, and order list

* added the config for the pagination for order, and products in the seller dashboard

najamulhassanarbi committed:
added reviews to products, along with permission (#8)

* added reviews to products, along with permission

* ready to deploy

najamulhassanarbi committed:
Reviews (#9)

* added reviews to products, along with permission

* ready to deploy

* requirement.txt

najamulhassanarbi committed:
Update build.sh

najamulhassanarbi committed:
Update settings.py

najamulhassanarbi committed:
added static files

najamulhassanarbi committed:
changed the pagination to hard coded value

najamulhassanarbi committed:
fixed cors issue

najamulhassanarbi committed:
completed the chat feature between seller and customer (#10)

* completed the chat feature between seller and customer

* here goes nothing

* here goes nothing

najamulhassanarbi committed:
added requirement

najamulhassanarbi committed:
Update requirements.txt

najamulhassanarbi committed:
we have to do it.
Merge branch 'ArbiShop' of https://github.com/najamulhassanarbi/BVStraining into ArbiShop

najamulhassanarbi committed:
added the setting file to django setting module

najamulhassanarbi committed:
changed the order of excution for django

najamulhassanarbi committed:
added the secur protocol

najamulhassanarbi committed:
added websocket libraris uvicorn and gunicorn
